torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,812 @@
+
+        # setting two factors to avoid underflow and overflow
+        self.gradient_predivide_factor: float = self._get_gradient_predivide_factor(
+            self.world_size
+        )
+        self.gradient_postdivide_factor: float = (
+            self.world_size / self.gradient_predivide_factor
+        )
+
+        self.numel_padded_per_param: List[int] = []
+
+        # Only handle params which are not already sharded. This enables
+        # sharding individual layers of a Module, with an outer wrapper to
+        # shard any leftover parameters.
+        params = []
+        for param_name, param in module.named_parameters():
+            if not hasattr(param, "_is_sharded"):
+                params.append(param)
+        self.params = params
+
+        # Shard module parameters in place
+        self._shard_parameters()
+
+        # Make sure all parameters are sharded.
+        for n, p in self.named_parameters():
+            if not hasattr(p, "_is_sharded"):
+                raise RuntimeError(f"found unsharded parameter: {n} ; {p.size()}")
+        self._reset_lazy_init()
+
+        # Enum to indicate if we're in the forward/backward pass, idle, etc.
+        self.training_state = TrainingState_.IDLE
+
+        # Flag to guard against preparing gradients multiple times per backward pass.
+        self._pre_backward_hook_has_run = False
+
+    # setting two factors 'self.gradient_predivide_factor'
+    # and 'self.gradient_postdivide_factor' to avoid underflow and overflow
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
+        return float(factor)
+
+    def _cast_buffers(
+        self, device: Optional[torch.device] = None, memo: Optional[Set] = None
+    ) -> None:
+        """Move all buffers to the given *device*.
+        If *device* are not given, then they will default to
+        ``self.compute_device``. In the
+        case of nested FSDP instances, we will respect the child instance's
+        ``compute_device`` configuration.
+        Args:
+            device (torch.device, Optional):
+                device to cast buffers to (defaults to compute_device)
+            memo (Set, Optional):
+                set of modules that have already been processed
+        """
+        if memo is None:
+            memo = set()
+        for module in self.modules():
+            if module is not self and isinstance(module, FullyShardedDataParallel):
+                # Allow any child FSDP instances to handle their own buffers.
+                module._cast_buffers(device=device, memo=memo)
+            elif module not in memo:
+                memo.add(module)
+                for name, buf in module.named_buffers(recurse=False):
+                    if buf is None:
+                        continue
+                    buf = buf.to(device=device or self.compute_device)
```

Comment:
Sometimes device is passed in as a number i.e. 0 to indicate cuda:0, is there a risk here because in this case it will defualt to self.compute_device?